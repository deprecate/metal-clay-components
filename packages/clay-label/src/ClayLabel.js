import 'clay-button';
import 'clay-link';
import Component from 'metal-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayLabel.soy.js';

/**
 * Metal Clay Label component.
 */
class ClayLabel extends Component {
  /**
	 * Handle `click` button and emit event `close`.
	 * @protected
	 */
  _handleCloseButtonClick() {
    this.emit('close');
  }
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayLabel.STATE = {
  /**
 	 * True or false to activate the close button.
 	 * @instance
 	 * @memberof ClayLabel
 	 * @type {?bool}
 	 * @default false
 	 */
  closeable: Config.bool().value(false),

  /**
 	 * HREF in label.
 	 * @instance
 	 * @memberof ClayLabel
 	 * @type {?string|undefined}
 	 * default undefined
 	 */
  href: Config.string(),

  /**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayLabel
	 * @type {?string|undefined}
	 * default undefined
	 */
  id: Config.string(),

  /**
	 * The label of the badge element.
	 * @instance
	 * @memberof ClayLabel
	 * @type {?string|undefined}
	 * default undefined
	 */
  label: Config.string(),

  /**
 	 * The size of the label element.
 	 * @instance
 	 * @memberof ClayLabel
 	 * @type {?string|undefined}
 	 * @default undefined
 	 */
  size: Config.oneOf(['sm']),

  /**
 	 * The path to the SVG spritemap file containing the icons.
 	 * @instance
 	 * @memberof ClayLabel
 	 * @type {?string|undefined}
 	 * @default undefined
 	 */
  spritemap: Config.string(),

  /**
	 * Label style color.
	 * @instance
	 * @memberof ClayLabel
	 * @type {?string}
	 * default secondary
	 */
  style: Config.oneOf([
    'danger',
    'info',
    'secondary',
    'success',
    'warning',
  ]).value('secondary'),
};

Soy.register(ClayLabel, templates);

export {ClayLabel};
export default ClayLabel;
