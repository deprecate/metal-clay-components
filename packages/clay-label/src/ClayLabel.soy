{namespace ClayLabel}

/**
 * This renders a label.
 */
{template .render}
	{@param? _handleCloseButtonClick: any}
	{@param? closeable: bool}
	{@param? href: string}
	{@param? id: string}
	{@param? label: string}
	{@param? size: string}
	{@param? spritemap: string}
	{@param? style: string}

	{let $classes kind="text"}
		label
		{sp}label-{$style ?: 'secondary'}

		{if $closeable}
			{sp}label-dismissible
		{/if}

		{if $size}
			{sp}label-{$size}
		{/if}
	{/let}

	<span class="{$classes}" id="{$id}">
		{call .content}
			{param _handleCloseButtonClick: $_handleCloseButtonClick /}
			{param closeable: $closeable /}
			{param href: $href /}
			{param label: $label /}
			{param spritemap: $spritemap /}
		{/call}
	</span>
{/template}

/**
 * Render icon and label content.
 */
{template .content}
	{@param? _handleCloseButtonClick: any}
	{@param? closeable: bool}
	{@param? href: string}
	{@param? label: string}
	{@param? spritemap: string}

	{if $href}
		{call ClayLink.render}
			{param href: $href /}
			{param label: $label /}
		{/call}
	{else}
		{$label}
	{/if}

	{if $closeable and $spritemap}
		{call ClayButton.render}
			{param ariaLabel: 'Close' /}
			{param elementClasses: 'close' /}
			{param events: [ 'click': $_handleCloseButtonClick] /}
			{param icon: [
				'alignment': 'right',
				'monospaced': false,
				'spritemap': $spritemap,
				'symbol': 'times'
			] /}
			{param style: 'unstyled' /}
			{param type: 'button' /}
		{/call}
	{/if}
{/template}
