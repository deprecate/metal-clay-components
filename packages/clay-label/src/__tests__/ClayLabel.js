'use strict';

import ClayLabel from '../ClayLabel';

let label;

describe('ClayLabel', function() {
  afterEach(() => {
    if (label) {
      label.dispose();
    }
  });

  it('shound generate the default markup', () => {
    label = new ClayLabel();

    expect(label).toMatchSnapshot();
  });

  it('shound render a label with label', () => {
    label = new ClayLabel({
      label: 'Foo',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a label with id', () => {
    label = new ClayLabel({
      id: 'Foo',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a large label', () => {
    label = new ClayLabel({
      size: 'lg',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a label with removable', () => {
    label = new ClayLabel({
      label: 'Foo',
      closeable: true,
      spritemap: 'icons.svg',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a label with link', () => {
    label = new ClayLabel({
      label: 'Foo',
      href: '#1',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a `default color` label', () => {
    label = new ClayLabel({
      style: 'secondary',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a `info color` label', () => {
    label = new ClayLabel({
      style: 'info',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a `warning color` label', () => {
    label = new ClayLabel({
      style: 'warning',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a `danger color` label', () => {
    label = new ClayLabel({
      style: 'danger',
    });

    expect(label).toMatchSnapshot();
  });

  it('shound render a `success color` label', () => {
    label = new ClayLabel({
      style: 'success',
    });

    expect(label).toMatchSnapshot();
  });
});
