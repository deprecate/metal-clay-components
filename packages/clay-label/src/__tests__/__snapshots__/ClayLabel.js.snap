// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayLabel shound generate the default markup 1`] = `<span class="label label-secondary"></span>`;

exports[`ClayLabel shound render a \`danger color\` label 1`] = `<span class="label label-danger"></span>`;

exports[`ClayLabel shound render a \`default color\` label 1`] = `<span class="label label-secondary"></span>`;

exports[`ClayLabel shound render a \`info color\` label 1`] = `<span class="label label-info"></span>`;

exports[`ClayLabel shound render a \`success color\` label 1`] = `<span class="label label-success"></span>`;

exports[`ClayLabel shound render a \`warning color\` label 1`] = `<span class="label label-warning"></span>`;

exports[`ClayLabel shound render a label with classes 1`] = `<span class="label label-secondary my-custom-class"></span>`;

exports[`ClayLabel shound render a label with id 1`] = `<span class="label label-secondary" id="myId"></span>`;

exports[`ClayLabel shound render a label with label 1`] = `<span class="label label-secondary">Foo</span>`;

exports[`ClayLabel shound render a label with link 1`] = `
<span class="label label-secondary">
  <a class="link-secondary" href="#1">Foo</a>
</span>
`;

exports[`ClayLabel shound render a label with removable 1`] = `
<span class="label label-secondary label-dismissible">Foo
  <button class="btn close btn-unstyled" aria-label="Close" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
      <use xlink:href="icons.svg#times"></use>
    </svg>
  </button>
</span>
`;

exports[`ClayLabel shound render a large label 1`] = `<span class="label label-secondary label-lg"></span>`;
